File: codegen/codegen-erlang/src/test/resources/ir/sigv4_helpers.expected.erl
resolve_host(#http_request{host = Host, headers = Headers}, Opts) ->
    coalesce([
        Host,
        maps:get(host, Opts, undefined),
        maps:get(endpoint_host, Opts, undefined),
        header_host(Headers)
    ]).

coalesce([H | Rest]) ->
    case H of
        undefined -> coalesce(Rest);
        <<>> -> coalesce(Rest);
        Value -> Value
    end;
coalesce([]) ->
    <<"localhost">>.

build_url(Host, Path, Query) ->
    <<"https://", Host/binary, Path/binary, (query_suffix(Query))/binary>>.

query_suffix(Query) when map_size(Query) =:= 0 -> <<>>;
query_suffix(Query) ->
    Params = uri_string:compose_query([{K, V} || {K, V} <- maps:to_list(Query)]),
    <<"?", Params/binary>>.

ensure_host_header(Headers, Host) ->
    case header_host(Headers) of
        undefined -> [{<<"host">>, Host} | Headers];
        _ -> Headers
    end.

header_host(Headers) ->
    proplists:get_value(<<"host">>, Headers, proplists:get_value(<<"Host">>, Headers)).

maybe_add_session_token(Headers, undefined) ->
    Headers;
maybe_add_session_token(Headers, Token) ->
    case proplists:get_value(<<"x-amz-security-token">>, Headers) of
        undefined -> [{<<"x-amz-security-token">>, Token} | Headers];
        _ -> Headers
    end.

sign_options(Service, Opts) ->
    [{uri_encode_path, Service =/= <<"s3">>}] ++ body_digest_option(Opts).

body_digest_option(Opts) ->
    case maps:get(unsigned_payload, Opts, false) of
        true -> [{body_digest, <<"UNSIGNED-PAYLOAD">>}];
        false -> []
    end.

session_token_option(undefined) -> [];
session_token_option(Token) -> [{session_token, Token}].

endpoint_host_from_config(Config) ->
    case maps:get(base_url, Config, undefined) of
        undefined ->
            case
                {
                    maps:get(endpoint_prefix, Config, undefined),
                    maps:get(region, Config, <<"us-east-1">>)
                }
            of
                {undefined, _} -> undefined;
                {Prefix, Region} -> <<Prefix/binary, ".", Region/binary, ".amazonaws.com">>
            end;
        BaseUrl ->
            {_Scheme, Authority} = split_base_url(BaseUrl),
            Authority
    end.

split_base_url(<<>>) ->
    {<<>>, <<>>};
split_base_url(BaseUrl) ->
    case uri_string:parse(binary_to_list(BaseUrl)) of
        #{scheme := Scheme, host := Host} = Parts ->
            PortSuffix =
                case maps:get(port, Parts, undefined) of
                    undefined -> <<>>;
                    Port -> <<":", (integer_to_binary(Port))/binary>>
                end,
            {<<(list_to_binary(Scheme))/binary, "://">>, <<
                (list_to_binary(Host))/binary, PortSuffix/binary
            >>};
        _ ->
            {<<>>, BaseUrl}
    end.

File: codegen/codegen-elixir/src/test/resources/ir/sigv4_helpers.expected.ex
def endpoint_host_from_config(config) do
  case Map.get(config, :base_url) do
    nil ->
      case {Map.get(config, :endpoint_prefix), Map.get(config, :region, "us-east-1")} do
        {nil, _} -> nil
        {prefix, region} -> "#{prefix}.#{region}.amazonaws.com"
      end
    base_url ->
      {_scheme, authority} = split_base_url(base_url)
      authority
  end
end

defp resolve_host(request = %HttpRequest{}, opts) do
  coalesce([request.host, Map.get(opts, :host), Map.get(opts, :endpoint_host), header_host(request.headers)])
end

defp coalesce([nil | rest]), do: coalesce(rest)
defp coalesce(["" | rest]), do: coalesce(rest)
defp coalesce([value | _]), do: value
defp coalesce([]), do: "localhost"

defp build_url(host, path, query) when map_size(query) == 0, do: "https://#{host}#{path}"
defp build_url(host, path, query) do
  params = URI.encode_query(query)
  "https://#{host}#{path}?#{params}"
end

defp ensure_host_header(headers, host) do
  case header_host(headers) do
    nil -> [{"host", host} | headers]
    _ -> headers
  end
end

defp header_host(headers) do
  Enum.find_value(
    headers,
    fn
      {"host", value} ->
        value;
      {"Host", value} ->
        value;
      _ ->
        nil
    end
  )
end

defp maybe_add_session_token(headers, nil), do: headers
defp maybe_add_session_token(headers, token) do
  if Enum.any?(headers, fn {k, _} -> String.downcase(k) == "x-amz-security-token" end) do
    headers
  else
    [{"x-amz-security-token", token} | headers]
  end
end

defp body_digest_option(opts) do
  if Map.get(opts, :unsigned_payload, false) do
    [{:body_digest, "UNSIGNED-PAYLOAD"}]
  else
    []
  end
end

defp session_token_option(nil), do: []
defp session_token_option(token), do: [{:session_token, to_bin(token)}]

defp split_base_url(""), do: {"", ""}
defp split_base_url(base_url) do
  case URI.parse(base_url) do
    %URI{scheme: scheme, host: host} = uri when is_binary(host) ->
      port_suffix =
        case uri.port do
          nil -> ""
          port -> ":#{port}"
        end
  
      {"#{scheme}://", "#{host}#{port_suffix}"}
  
    _ ->
      {"", base_url}
  end
end

defp to_bin(value) when is_binary(value), do: value
defp to_bin(value) when is_atom(value), do: Atom.to_string(value)
defp to_bin(value), do: Kernel.to_string(value)

defp to_erl_headers(headers) do
  Enum.map(headers, fn {k, v} -> {to_bin(k), to_bin(v)} end)
end

defp from_erl_headers(headers) do
  Enum.map(headers, fn {k, v} -> {Kernel.to_string(k), Kernel.to_string(v)} end)
end